Consolidate GPU Kernel launches

spencerw opened the pull request:
Having GPU kernel launches tied to the TreePieces degrades performance and is probably causing some data race conditions. The GPU versions of the local tree walk and ewald calculation are now handled by the data manager. Any kernel launches involving interaction list calculations are handled by node groups.

Note that another open PR #183 has already been merged into this branch.

spencerw committed:
clearRegisteredPieces called before tree build

spencerw committed:
CUDA streams assigned to TreePieces after tree build

spencerw committed:
Move clearRegisteredPieces inside buildTree

spencerw committed:
Remove unused code

spencerw committed:
Move assignCUDAStreams() call inside of buildTree()

spencerw committed:
Remove DataTransfer calls from TreePieceCellListDataTransferLocal when GPU tree walk enabled

spencerw committed:
Hack to launch gpuLocalTreeWalk from DataManager

spencerw committed:
Hack to run remote gravity without bookkeeping

spencerw committed:
Restore bookkeeping around GPU local tree walk

spencerw committed:
Use bare callback for local walk

spencerw committed:
First attempt at consolidated Ewald GPU kernel

spencerw committed:
Merge remote-tracking branch 'remotes/origin/dm_tp' into kernelfix

spencerw committed:
More fixes to consolidated Ewald

spencerw committed:
Ensure device memory pointers are passed to TreePieces even though commenceCalculateGraviyLocal is no longer called

spencerw committed:
Call finishBucket after local tree walk, ensure smooth happens before Ewald

spencerw committed:
Pass nReplicas and fPeriod to DataManagerLocalTreeWalk

spencerw committed:
Removed unused Ewald GPU code

spencerw committed:
Ewald GPU data to pinned host memory, remove markers

spencerw committed:
Comments and code cleanup

spencerw commented:
We still need to decide what to do about the nodeGravityComputation and particleGravityComputation kernel launches. I don't think the remote gravity performance will benefit much from consolidating these, but this PR as-is probably breaks the local GPU gravity calculation if we aren't using the gpu-local-tree-walk option.

trquinn commented:
This doesn't even compile if "--enable-gpu-local-tree-walk" is not specified.

spencerw committed:
Fix for bucket bookkeeping

spencerw commented:
I just tested this out using the verbs comm layer and CUDA memory errors are back. I'm guessing the poor performance from MPI was actually preventing the remote gravity kernels from stepping on each other.

We'll see if the CSA folks have any other suggestions when we talk to them next Monday, but I think we're going to need to move all of the nodeGravityComputation and particleGravityComputation kernel launches to the DataManager as well.

spencerw committed:
Temporarily move initiatePrefetch to commenceCalculateGravityLocal

spencerw committed:
Interaction list kernels launch once per PE

spencerw committed:
Bug fixes for Ewald and multistepping

spencerw committed:
GPU local walk and Ewald callbacks use treeProxy

spencerw committed:
Remove small phase code

spencerw committed:
Remove CUDA_2D_FLAT option

spencerw committed:
Use HAPI callbacks for PEList functions

mrcawood committed:
dMProxy mem log writer

mrcawood committed:
fixed headers

spencerw committed:
CUDA streams handled by PE groups

mrcawood committed:
fixed init, added nodeID

spencerw committed:
PEList vectors use pinned host memory

mrcawood committed:
merging into kernelfix

spencerw committed:
ifdefs around GPU code

mrcawood committed:
added origin function tag

spencerw committed:
Fix DataManager creating CUDA stream too early

spencerw committed:
Callbacks for PEList functions managed by class

spencerw committed:
Fix memory leaks

spencerw committed:
Various code cleanup

* Callback objects managed by PEList
* PEList::setType handled by constructor

mrcawood committed:
updates for memory logger

spencerw committed:
Merge branch 'kernelfix' into kernelfix

spencerw committed:
Fixes to Ewald on GPU

* Remove check for active particles in EwaldGPU
* Add HAPI callback after EwaldKernel launch

spencerw committed:
Merge pull request #3 from mrcawood/kernelfix

memory usage writer

spencerw committed:
Cleanup for GPU memory logger

spencerw committed:
Remove HAPI_MEMPOOL config option

spencerw committed:
Fixes to CUDA build options

*Remove CUDA_PRINT_ERRORS option
*Proper placement of CUDA_NO_KERNELS macros
*Add CUDA_VERBOSE_OPS option
*Add new HAPI_TRACE markers

spencerw committed:
Comments and docstrings

spencerw committed:
Fix hang when a PE has no TreePieces

spencerw committed:
Performance tweaks

* Wait until finishLocalWalk to free pinned host memory
* Fix missing user event registration for CUDA_GRAV_TREE_LOCAL

spencerw committed:
Move initiatePrefetch back to startGravity

* Prevent prefetch data transfer to GPU until local data transfer
  completes
* Dont use pinned host memory for remote gravity interaction lists

spencerw committed:
Fix PINNED_HOST_MEMORY option

trquinn reviewed (changes requested):
Mostly cleanups, but I think there are a couple things to discuss.

spencerw commented on State.h:
```diff
@@ -168,6 +168,7 @@ class DoubleWalkState : public State {
   std::unordered_map<NodeKey,int> nodeMap;
   std::unordered_map<NodeKey,int> partMap;
 
+  // TODO do these need to be shut off (no restriction)?
```
From what I can tell, these functions are only used during the remote-resume phase. Since we now wait for the RR to finish before shipping anything to the GPU, I don't think these checks will actually save us from running out of memory.

spencerw committed:
Fixes according to Tom's PR comments

* Add array to iterate over PEList groups
* Separate CkCallback objects for different GPU phases
* Remove CkVec from PEList class
* More comments and documentation
* Compute doesnt put missedNodes and missedParts in pinned host memory
* Remove createStream function
* Remove malloc after packacing up remote-resume CUDA request
* Remove unused h_idata

spencerw committed:
Free remote buffers in resumeRemoteChunk

spencerw commented on DataManager.cpp:
```diff
@@ -500,33 +486,151 @@ void DataManager::serializeLocalTree(){
+  }
+
+  if (registeredTreePieces[0].treePiece->bEwald) {
+    allocatePinnedHostMemory((void **)&h_idata, sizeof(EwaldData)*savedNumTotalParticles-1);
```
h_idata has an EwaldRange field, which was used to store the particle indices back when EwaldKernel was launched per TreePiece. This is no longer needed. Besides that, it stored a pointer to 'cachedData' and 'ewt', which are now handled by the DataManager.

I agree that all references to this can be removed, along with the 'EwaldData' struct.

spencerw commented on Ewald.cpp:
```diff
@@ -383,64 +383,53 @@ void TreePiece::EwaldInit()
-  ewtTable = (EwtData*) h_idata->ewt;
-  roData = (EwaldReadOnlyData*) h_idata->cachedData; 
+
+		ewtTable = (EwtData*)_ewtTable;
```
I think I can move all of the ewtTable and cachedData code to the DataManager. However, I see that nEwhLoop is needed, which appears to get set  in TreePiece::EwaldInit. It looks like this is set dynamically by each TreePiece, but I'm guessing it should be the same everywhere. Is there a way to derive this from constants or simulation parameters?

spencerw committed:
Move TreePiece::EwaldGPU code to DataManager::startEwald

spencerw committed:
Ensure Ewald still triggers if GPU_LOCAL_TREE_WALK is disabled

spencerw committed:
Various minor fixes

spencerw committed:
Add back in PinnedHostAllocator for PEList

mrcawood committed:
integrated memory pools \w warmup back into upstream kernelfix

mrcawood committed:
increased host pool warmup buckets

mrcawood committed:
added logging to CPU memory pool with +cpumemlog

mrcawood committed:
removed large long lived mallocs from host mem pool

mrcawood committed:
Many improvements for host memory pool, include trimming and refill. Added stats ouput. See MEMORY_POOL_CHANGES.txt for more info

spencerw committed:
Allow remote walk before GPU data transfer

* Remove barrier preventing calculateGravityRemote from executing before
  prefetch data is transferred to GPU
* Add barrier ensuring prefetch data is transferred before launching
  PEList kernels
* Node-wide GPU data pointers no longer passed through TreePieces

mrcawood committed:
removed large mallocs from pool, further tuned warmup and trim params

spencerw commented on DataManager.cpp:
```diff
@@ -702,28 +700,11 @@ void DataManager::resumeRemoteChunk() {
+ // Check and see if the remote walks already finished and are waiting
+ // to launch their GPU kernels
+ for (int i = 0; i < numPEListProxies; i++) {
+     PEListProxies[i]->tryLaunchDelayedKernel();
```
This doesn't seem to be doing what I want. There are 6 PEList proxies, so I need 6 invocations of this function on each PE.

When I run on a single node with 2 ranks, I'm seeing this function get called 12 times on each PE. Any idea what's going on here? Is there a better way to do this?

spencerw replied:
I was wondering if this was possible. I imagine you mean something like

```
 int pe;
 int firstPE = CkNodeFirst(CkMyNode());
 int nPEs = CkNodeSize(CkMyNode());
 for (int i = 0; i < numPEListProxies; i++) {
     for (int j = 0; j < nPEs; j++) {
         pe = firstPE + j;
         PEListProxies[i][pe].tryLaunchDelayedKernel();
     }
 }
```

The function call on the proxy is giving me a seg fault here. It's not clear to me if ordering is guaranteed here.

spencerw replied:
The other way I was thinking of doing this was to maintain a 'registeredPEList' array in the DataManager. Similar to the 'registeredTreePieces', I'm wondering if the 'thisIndex' property is necessary to iterate over the elements of the PE group like this.

spencerw replied:
Oh right, I forgot that I made PEListProxies an array of pointers. The intention there was to make the sendList() calls in Compute.cpp easier to read.

Your suggestion appears to work. Although I still haven't seen any cases where the remote walk beats the remote data transfer and triggers the protection against this race condition.

spencerw committed:
resumeRemoteChunk iterates over PEList proxies correctly

spencerw committed:
fillGPUBuffer given priority over SPH

mrcawood committed:
removed stale code

mrcawood committed:
Merge remote-tracking branch 'upstream/kernelfix' into kernelfix

mrcawood committed:
Fix thread safety: synchronize shared data access and prevent deadlocks in host memory pool

spencerw committed:
* Fix hang when an entire PE has no particles
* CUDA stream for DataManager created later

spencerw committed:
Merge remote-tracking branch 'upstream/master' into kernelfix

mrcawood committed:
Fix GPU memory pool race condition causing non-determinism

cudaFreeAsync returns memory to CUDA's pool immediately, but GPU operations
using that memory may still be in flight. When cudaMallocAsync returns the
same memory, there's a race condition causing non-deterministic results.

Fix: Synchronize the stream after cudaFreeAsync to ensure all operations
complete before the memory becomes available for reuse.

This fixes the teststep/testenergy benchmark which was failing due to
~0.04% energy calculation errors from the race condition.

spencerw committed:
Merge remote-tracking branch 'matt/kernelfix' into kernelfix

mrcawood committed:
added more null pointer guardrails

spencerw committed:
Rework PEList callbacks

spencerw committed:
Remove pinned host allocator

spencerw committed:
Fix minor memory leaks

spencerw committed:
Remove stream init from DataManager constructor

spencerw committed:
Merge pull request #5 from mrcawood/kernelfix

Host memory pool fully featured

spencerw commented:
I went ahead and merged Matt's memory pool PR into this. As of that latest commit (04724af), teststep, testcosmo and testcollapse all appear to pass and are coming back clean with valgrind, address-sanitizer and compute-sanitizer when run on multiple ranks.

We've also confirmed that the halo mass function matches between the CPU and GPU version of runs22, although this was run a couple months ago with an older commit (10ed833).

At the point, the main issue is that the star formation history (using metal cooling) for CosmoRun diverges between the CPU and GPU version somewhere around z=7. I'm also getting a heap corruption error around this time. This was all run with the verbs build on Vista. I also had to roll back to (10ed833), otherwise CosmoRun crashes and hangs fairly quickly. The MPI/UCX build works much more smoothly with the latest commit (04724af), but I run into RDMA errors 20 or so steps in.

```
[c637-132:2942276:0:2942276] ib_mlx5_log.c:179  Local length error on mlx5_0:1/IB (synd 0x1 vend 0xda hw_synd 0/0)                   [c637-132:2942276:0:2942276] ib_mlx5_log.c:179  RC QP 0x3db8 wqe[58529]: RDMA_READ s-- [rva 0x40034f29cd60 rkey 0x9b2a95c] [va 0x4f5d5e30 len 16384 lkey 0x15accae0] [rqpn 0x10c45 dlid=734 sl=0 port=1 src_path_bits=0]
==== backtrace (tid:2942276) ====
 0  /opt/apps/ucx/1.17.0/lib/libucs.so.0(ucs_handle_error+0x288) [0x40002030eb98]
 1  /opt/apps/ucx/1.17.0/lib/libucs.so.0(ucs_fatal_error_message+0xd0) [0x40002030c1e0]
 2  /opt/apps/ucx/1.17.0/lib/libucs.so.0(ucs_log_default_handler+0xd80) [0x4000203105f0]
 3  /opt/apps/ucx/1.17.0/lib/libucs.so.0(ucs_log_dispatch+0xd0) [0x400020310950]
 4  /opt/apps/ucx/1.17.0/lib/ucx/libuct_ib.so.0(uct_ib_mlx5_completion_with_err+0x420) [0x4000225e6b50]
 5  /opt/apps/ucx/1.17.0/lib/ucx/libuct_ib.so.0(+0x3d694) [0x4000225fd694]
 6  /opt/apps/ucx/1.17.0/lib/ucx/libuct_ib.so.0(uct_ib_mlx5_check_completion+0x68) [0x4000225e7a9c]
 7  /opt/apps/ucx/1.17.0/lib/ucx/libuct_ib.so.0(+0x3f364) [0x4000225ff364]
 8  /opt/apps/ucx/1.17.0/lib/libucp.so.0(ucp_worker_progress+0x64) [0x4000201d8ab8]
 9  /opt/apps/nvidia24/openmpi/5.0.5_nvc249/lib/libmpi.so.40(mca_pml_ucx_recv+0x134) [0x40001f178034]
10  /opt/apps/nvidia24/openmpi/5.0.5_nvc249/lib/libmpi.so.40(PMPI_Recv+0x1bc) [0x40001f00053c]
11  /work/03777/scw7/vista/changa_kernelfix_group_matt/ChaNGa.smp.cuda() [0xa97660]
12  /work/03777/scw7/vista/changa_kernelfix_group_matt/ChaNGa.smp.cuda(_Z24LrtsAdvanceCommunicationi+0x648) [0xa97fc8]
13  /work/03777/scw7/vista/changa_kernelfix_group_matt/ChaNGa.smp.cuda(_Z25CommunicationServerThreadi+0x1c) [0xa94e1c]
14  /work/03777/scw7/vista/changa_kernelfix_group_matt/ChaNGa.smp.cuda() [0xa94cac]
15  /work/03777/scw7/vista/changa_kernelfix_group_matt/ChaNGa.smp.cuda(ConverseInit+0x1694) [0xa94454]
16  /work/03777/scw7/vista/changa_kernelfix_group_matt/ChaNGa.smp.cuda(charm_main+0x2c) [0xa0566c]
17  /lib64/libc.so.6(+0x2c79c) [0x40001f9ec79c]
18  /lib64/libc.so.6(__libc_start_main+0x98) [0x40001f9ec86c]
19  /work/03777/scw7/vista/changa_kernelfix_group_matt/ChaNGa.smp.cuda(_start+0x30) [0x5cadb0]
=================================
```

Disabling the UCX transport layer with `export OMPI_MCA_btl=^uct` seems to help it get a bit further, but it still crashes with the same error eventually.

h896 seems much more stable, although I haven't run it for more than about 50 steps.

trquinn commented:
The GPU code currently hangs if running on one node.  I believe it is because there are no remote walks so DataManager::transferParticleVarsBack() is not called enough times.

trquinn committed:
TreePiece::finishWalk(): make HAPI_TRACE less verbose.

trquinn committed:
PEList and DataManager: add booleans to track data transfer to the GPU.

Previously this was done using the device pointers, which occasionally
caused hangs if no data was transfered to the GPU.  (E.g. a remote walk
when running on a single node.)

Also cleaned up some unused attributes in DataManager.

trquinn commented:
> This doesn't even compile if "--enable-gpu-local-tree-walk" is not specified.

I've checked that it seems to run correctly with or without `--enable-gpu-local-tree-walk` after my recent patch (yet to be merged).

trquinn committed:
Synchronously call EwaldInit() to avoid use-before-init problem.

trquinn committed:
PEList::launch_kernel(): remove asserts that are not valid.
If there are no remote walks, these asserts will fail.

mrcawood committed:
Fix SMP memory visibility in remote gravity data transfer gating

Replace plain bool with std::atomic<bool> for bLocalDataTransferred and
bRemoteDataTransferred. These flags are set in async callbacks and read by
multiple PEs in PEList::finishWalk(); without atomics, some PEs see stale
values and never set bKernelDelayed, causing tryLaunchDelayedKernel to
run before they check in → deadlock.

- DataManager: .store() for all writes; .load() in PEList when reading
- donePrefetch path also sets bRemoteDataTransferred via .store(true)

mrcawood committed:
Host pool init race fix + MemoryPool review cleanups

* fixed host pool init race condition

* Address PR #7 review: MemoryPool.cpp cleanups

- gpuPoolInit: Replace magic 16 with const maxCudaDevice; add cudaGetDeviceCount
  and CkAssert(nCudaDeviceCount <= maxCudaDevice)
- gpuPoolInit: Check cudaMemPoolSetAttribute return value, CkAbort on failure
- hostPoolReportStats: Replace snprintf with make_formatted_string per project
  pattern (formatted_string.h)

mrcawood committed:
Fix startEwaldGPU crash when no TreePiece has valid Ewald data (#13)

* Fix startEwaldGPU crash when no TreePiece has valid Ewald data

Previously used registeredTreePieces[0] blindly. When no TP has root/ewt/
nEwhLoop valid (e.g. after migration), this caused crashes. Now search for
first TP with valid Ewald data and CkAbort with clear message if none found.
Add nEwhLoop <= NEWH check. NEWH 80→400 to avoid exceed for some runs.

* PR #13: Clarify cudaFinishAllBuckets(1) and valid-TP search (trquinn review)

- Explain why we call cudaFinishAllBuckets(1) in early-exit: finishEwaldGPU()
  (the Ewald kernel callback) never runs when we skip the kernel; we replicate
  it here to avoid deadlock.
- Tighten valid-TP comment: registeredTreePieces[0] may be empty or lack data.

mrcawood committed:
Allocate particleInterRemote before finishWalk to avoid NULL dereference in callbacks (#14)

* Allocate particleInterRemote early in startGravity for migration safety

PUP sets particleInterRemote to NULL on unpack. Allocation was after
myNumParticles==0 early return; callbacks could arrive before allocation.
Move allocation before early return. Add defensive NULL checks in Compute
paths (recvdParticles, nodeRecvdEvent) that use particleInterRemote.

* PR #14: Address trquinn review - CkAssert for particleInterRemote, clarify TreePiece comment

trquinn committed:
Inactive TreePiece optimization (#18)

* PEList::launch_kernel(): remove asserts that are not valid.
If there are no remote walks, these asserts will fail.

* TreePiece::startGravity(): optimize for 0 active particles

TreePiece::myNumActiveParticles are now available for non-GPU code.

TreePiece::initBuckets() now calculates the number of active particles.

DataManager calls to TreePiece::cudaFinishAllBuckets(),
commenceLocalGravity() and continueStartRemoteChunk() now all have to
check for active particles to avoid accessing unallocated
Compute/Walk/State objects.

mrcawood committed:
Fix PEList launch readiness when local transfer lags remote transfer (#19)

Require bLocalDataTransferred for every PEList launch; for remote also require bRemoteDataTransferred. Re-check readiness in tryLaunchDelayedKernel; trigger delayed-kernel retry from DataManager::startLocalWalk as well as resumeRemoteChunk.

trquinn committed:
class PEList: remove unused attribute.

trquinn committed:
TreePiece::updateParticles(): skip inactive TreePieces.

This is needed to avoid crashes caused by PR#18.

trquinn committed:
TreePiece constructors: initialize nEwhLoop.

nEwhLoop is checked for a valid value in DataManager before loading
the Ewald h table onto the GPU.  It needs to be set to an invalid
value before the h-table is constructed.

trquinn committed:
White space cleanup.

trquinn committed:
TreePiece::startGravity(): empty TPs don't checkin with PEList::finishWalk().

PELists only work with non-empty tree pieces.

Also add CkAssert() in PEList::finishWalk() to catch too many checkings
from TPs.

trquinn commented on DataManager.cpp:
```diff
@@ -500,33 +486,151 @@ void DataManager::serializeLocalTree(){
+      if(registeredTreePieces[0].treePiece->bEwald) {
+        EwaldMsg *msg = new (8*sizeof(int)) EwaldMsg;
+        msg->fromInit = false;
+        // Make priority lower than smooth.
```
Outdated.

trquinn committed:
PEList: change TP checkin counter from vector to an int.

Using a vector of TreePiece pointers was overkill since we just needed a
counter.

Also added needed (for checkpoint/restart) code in Migration constructor and pup.

trquinn committed:
ListCompute::stateReady(): deleted unused variables.

trquinn committed:
class TreePiece: consolidate numActiveParticles and myNumActiveParticles

These two attributes kept track of the same thing.
